[PVR] Allow custom genre on recordings

hackthis02 opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
Allows PVR plugins to set custom text for genre type the same way EPG is already doing.

## Motivation and Context
By default Kodi only supports a hand full of genre types in the PVR EPG there is already a workaround
by letting the PVR plugin use localized strings as genre. The functionality was needed in the PVR
recording screen.

## Screenshots (if appropriate):
[before](http://imgur.com/rXQCsSj)
[after](http://imgur.com/pPWPLHW)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

hackthis02 committed:
Allow custom genre on PVR recordings

Allow PVR plugins to set custom genre strings for recordings the same
way EPG can.

hackthis02 committed:
Allow custom genre on PVR recordings

Allow PVR plugins to set custom genre strings for recordings the same
way EPG can.

hackthis02 committed:
Fixed indentation

Fixed indentation

ksooo commented:
note: changing pvr addon api by inserting a new struct member is a breaking change, because all pvr addons must be at least recompiled. No breaking changes for krypton in beta cicle. Thus v18.

hackthis02 commented:
I didn't realize it would require a recompile across the board. Thanks for the heads up.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h:
```diff
@@ -460,6 +460,7 @@ extern "C" {
     unsigned int    iMarginEnd;                                /*!< @brief (optional) if set, the backend ends the recording iMarginEnd minutes after endTime. */
     int             iGenreType;                                /*!< @brief (optional) genre type */
     int             iGenreSubType;                             /*!< @brief (optional) genre sub type */
+    const char *    strGenreDescription;                       /*!< @brief (optional) genre. Will be used only when iGenreType = EPG_GENRE_USE_STRING */
```
This seems not to be used anywhere. So, either add an implementation to kodi pvr core or drop this line, please.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h:
```diff
@@ -487,6 +488,7 @@ extern "C" {
     int    iLifetime;                                     /*!< @brief (optional) life time in days of this recording */
     int    iGenreType;                                    /*!< @brief (optional) genre type */
     int    iGenreSubType;                                 /*!< @brief (optional) genre sub type */
+    const char *    strGenreDescription;                  /*!< @brief (optional) genre. Will be used only when iGenreType = EPG_GENRE_USE_STRING */
```
we use character buffers, not pointers to addon internal structure. please change. buffer size could be PVR_ADDON_DESC_STRING_LENGTH.

ksooo commented on xbmc/pvr/recordings/PVRRecording.cpp:
```diff
@@ -94,7 +94,7 @@ CPVRRecording::CPVRRecording(const PVR_RECORDING &recording, unsigned int iClien
   m_strStreamURL                   = recording.strStreamURL;
   m_strChannelName                 = recording.strChannelName;
-  m_genre                          = StringUtils::Split(CEpg::ConvertGenreIdToString(recording.iGenreType, recording.iGenreSubType), g_advancedSettings.m_videoItemSeparator);
+  m_genre                          = StringUtils::Split(CPVRRecording::SetGenre(recording.iGenreType, recording.iGenreSubType, recording.strGenreDescription), g_advancedSettings.m_videoItemSeparator);
```
Please remove <code>CPVRRecording::</code> here, because it is not needed.

ksooo commented on xbmc/pvr/recordings/PVRRecording.h:
```diff
@@ -237,6 +237,14 @@ namespace PVR
     std::string EpisodeName(void) const { return m_strShowTitle; };
 
+    /*!
+    * @brief Change the genre of this recording.
```
please describe in more detail what the member function actually does, esp. the fallback strategy.

ksooo commented on xbmc/pvr/recordings/PVRRecording.cpp:
```diff
@@ -481,3 +481,18 @@ int CPVRRecording::ClientID(void) const
+    /* Determine the genre description from the type and subtype IDs */
+    return CEpg::ConvertGenreIdToString(iGenreType, iGenreSubType);
+  }
+}
```
well, as this is a member function called "SetGenre" it actually should do that. please change return type to void and set m_genre directly here.

hackthis02 committed:
Changes suggested by Ksooo

Switched strGenreDescription from char pointer to char buffer.
Changed SetGenre to set m_genre directly.
Removed strGenreDescription from PVR_TIMER struct.

hackthis02 commented:
Made the changes you suggested. Please tell me if there are any other issue/concerns you have.

ksooo commented on xbmc/pvr/recordings/PVRRecording.h:
```diff
@@ -237,6 +237,15 @@ namespace PVR
     std::string EpisodeName(void) const { return m_strShowTitle; };
 
+    /*!
+    * @brief Set m_genre to plugin proved sting when iGenreType = EPG_GENRE_USE_STRING
```
let me make suggestion for the complete section. I would write it like this:
<pre>
    * @brief set the genre for this recording.
    * @param iGenreType The genre type ID. If set to EPG_GENRE_USE_STRING, set genre to the value provided with strGenre. Otherwise, compile the genre string from the values given by iGenreType and iGenreSubType
    * @param iGenreSubType The genre subtype ID
    * @param strGenre The genre
</pre>

ksooo commented on xbmc/pvr/recordings/PVRRecording.h:
```diff
@@ -237,6 +237,15 @@ namespace PVR
+    * @param iGenreSubType The genre subtype ID
+    * @param strGenre custom genre
+    */
+    void SetGenre( int iGenreType, int iGenreSubType, char strGenre);
```
char strGenre => const std::string &strGenre

ksooo commented on xbmc/pvr/recordings/PVRRecording.cpp:
```diff
@@ -104,6 +103,7 @@ CPVRRecording::CPVRRecording(const PVR_RECORDING &recording, unsigned int iClien
   m_bIsDeleted                     = recording.bIsDeleted;
   m_iEpgEventId                    = recording.iEpgEventId;
   m_iChannelUid                    = recording.iChannelUid;
+  SetGenre(recording.iGenreType, recording.iGenreSubType, *recording.strGenreDescription);
```
If you change paramter type like I suggested, no need for * in front of recording.strGenreDescription any more.

ksooo commented on xbmc/pvr/recordings/PVRRecording.cpp:
```diff
@@ -481,3 +481,18 @@ int CPVRRecording::ClientID(void) const
+  if ((iGenreType == EPG_GENRE_USE_STRING) && (strGenre != NULL) && (sizeof(strGenre) > 0))
+  {
+    /* Type and sub type are not given.
+    * Use the provided genre description as backup. */
```
is the description really a "backup"?

hackthis02 committed:
Additional suggestions from Ksooo

Changed parameters on SetGenre
Updated SetGenre details

hackthis02 commented:
Thanks for the suggestion. This is my first project of this type. I'm still getting familiar with a project of this size and scope.

ksooo commented:
> Thanks for the suggestion. This is my first project of this type. I'm still getting familiar with a project of this size and scope.

 No problem. I'm glad that i can help.

ksooo commented:
BTW, which pvr addon is going to implement this feature? Can you show me the addon code? I do not like to add features to the core which are not used by any addon.

hackthis02 commented:
I'm using Media portal. I haven't submitted a pull request for the changes yet. I was waiting on the core to have this pull before submitting that one.
 
[Here](https://github.com/hackthis02/pvr.mediaportal.tvserver/commit/5bbf7741e022361506c3c756ec9124afbe9f5c8f) is my fork commit. Also @margro was already setting m_genre_type to EPG_GENRE_USE_STRING if they couldn't find a match. I just set it up to keep the custom string for use later.

From what I could see all the other PVR addons were doing basically the same thing. If the genre didn't match Kodi's predefined ones it just displayed as "Other / Unknown". I found this odd since this functionality is already present in the EPG. It seems pretty straight forward on the side of the addon to implement. So I'm hopeful if you marge it they will all start using it.

hackthis02 commented:
BTW if @phil65  is taking out the genre in the PVR recording screen this PR isn't as big of a deal.

ksooo commented:
yep, can we close this PR?

hackthis02 commented:
I still believe it's a valid feature improvement as it will still help if the user looks at the details dialog of a PVR recording but if you feel that's not enough to warrant it, sure.

ksooo commented:
let's keep it open for now. plenty of time until master is open for v18 development...

da-anda commented:
v18 development is open now. time to bring this PR in shape if it's still wanted.

hackthis02 commented:
Thanks for the heads up. I'll update it this week.

hackthis02 committed:
Merge remote-tracking branch 'refs/remotes/xbmc/master'

hackthis02 committed:
Updated to current version of file.

hackthis02 committed:
Revert "Updated to current version of file."

This reverts commit 6c41c052dc4d26d258c4aa400b836a02eb79beed.

hackthis02 committed:
Merge branch 'master' into master

hackthis02 committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	xbmc/pvr/recordings/PVRRecording.cpp
#	xbmc/pvr/recordings/PVRRecording.h

ksooo commented:
@hackthis02 whatever you are currently doing with git, it makes not much sense. could you please stop it? Every of your git activities spams a lot of team kodi members (and others). thanks you.

hackthis02 commented:
Oh I was unaware. I'm very sorry. I'm still learning git (much more familiar with TFS). I was attempting to squash my changes. If that's not a prerequisite for PR on this repo, I'll leave it as is.

hackthis02 committed:
Removed left in merge data

da-anda commented:
feel free to do as many commits in your local repo as you like, but please don't sync them automatically with your branch on Github during development. Once you're comfortable with your changes, squash them in reasonable commits and only then update your Github branch.

hackthis02 commented:
My local branch got hosed. So I'm closing this PR and opened a new one PR [#11553](https://github.com/xbmc/xbmc/pull/11553)
